include/thread_pool/thread_pool.hpp

```diff
@@ -109,35 +106,33 @@ class GenericThreadPool final
     */
     size_t getWorkerId();
 
-    SlottedBag<Queue> m_idle_workers;
-    WorkerVector m_workers;
-    Rouser m_rouser;
     size_t m_failed_wakeup_retry_cap;
     std::atomic<size_t> m_next_worker;
-    std::atomic<size_t> m_num_busy_waiters;
+    std::shared_ptr<Rouser> m_rouser;
+    std::shared_ptr<ThreadPoolState<Task, Queue>> m_state;
+    std::atomic<bool> m_moved;
```

Comment:
I think you don't need an additional flag m_moved because:
1) Construction/assignment/destruction should be always synchronized and no other member function should be called asynchronously during construction/destruction. 
2) After the move m_state will be null and you can use that to check and throw if a member function is called on a moved thread pool.

Reply:
👍  will use the null check

Reply:
can we guarantee atomicity and memory ordering if we rely on the shared pointer being null after `move` to check whether or not we can make a new post?